### PR TITLE
Make `Prover` generic over the `ConstraintCommitment` type

### PR DESCRIPTION
**Pull request by gswirski**
For better GPU performance, allow provers to customize `ConstraintCommitment` implementation/memory layout.

Implementors of the `Prover` trait would need to add the following lines since default associated types are unstable.
```
    type ConstraintCommitment<E: FieldElement<BaseField = Felt>> = DefaultConstraintCommitment<E, H, Self::VC>;

   [...]

    #[maybe_async]
    fn build_constraint_commitment<E: FieldElement<BaseField = Felt>>(
        &self,
        composition_poly_trace: CompositionPolyTrace<E>,
        num_constraint_composition_columns: usize,
        domain: &StarkDomain<Self::BaseField>,
        partition_options: PartitionOptions,
    ) -> (Self::ConstraintCommitment<E>, CompositionPoly<E>) {
        DefaultConstraintCommitment::new(composition_poly_trace, num_constraint_composition_columns, domain, partition_options)
    }
```

### CODE REVIEW

File: prover/src/constraints/composition_poly.rs [hunk 1]
```diff
@@ -48,7 +48,7 @@ impl<E: FieldElement> CompositionPolyTrace<E> {
 /// For example, if the composition polynomial has degree 2N - 1, where N is the trace length,
 /// it will be stored as two columns of size N (each of degree N - 1).
 pub struct CompositionPoly<E: FieldElement> {
-    data: ColMatrix<E>,
+    pub data: ColMatrix<E>,
```
Comment (gswirski) [inline prover/src/constraints/composition_poly.rs:51]: This should be reverted.

File: prover/src/matrix/col_matrix.rs [hunk 2]
```diff
@@ -217,6 +217,19 @@ impl<E: FieldElement> ColMatrix<E> {
         self
     }
 
+    pub fn evaluate_columns(&self) -> Self {
```
Comment (gswirski) [inline prover/src/matrix/col_matrix.rs:220]: This might be unnecessary.

File: prover/src/lib.rs [hunk 3]
```diff
@@ -150,6 +148,11 @@ pub trait Prover {
     where
         E: FieldElement<BaseField = Self::BaseField>;
 
+    /// Trace low-degree extension for building the LDEs of trace segments and their commitments.
+    type ConstraintCommitment<E>: ConstraintCommitment<E, HashFn = Self::HashFn, VC = Self::VC>
+    where
+        E: FieldElement<BaseField = Self::BaseField>;
```
Comment (irakliyk) [inline prover/src/lib.rs:154]: The comment here needs to be updated (I think it is just copy over from the `TraceLde` comment above).

File: prover/src/lib.rs [hunk 4]
```diff
@@ -524,45 +527,10 @@ pub trait Prover {
...
-        )
-        .in_scope(|| {
-            let commitment = composed_evaluations
-                .commit_to_rows::<Self::HashFn, Self::VC>(self.options().partition_options());
-            ConstraintCommitment::new(composed_evaluations, commitment)
-        });
-
-        (constraint_commitment, composition_poly)
-    }
+        E: FieldElement<BaseField = Self::BaseField>;
```
Comment (irakliyk) [inline prover/src/lib.rs:533]: I would move this method into the "Required methods" section (i.e., after line 204).

**Review by irakliyk — APPROVED** [review]
Looks great! Thank you! I left a couple of small comments inline. Also, we should update the main [README](https://github.com/facebook/winterfell/blob/main/README.md) file (as it contains an example of how to instantiate a prover (i.e., `WorkProver`)) and the `prover` crate [README](https://github.com/facebook/winterfell/blob/main/prover/README.md#prover) file.

**Review by irakliyk — APPROVED** [review]
All looks good! Thank you!

**Pull request merged**